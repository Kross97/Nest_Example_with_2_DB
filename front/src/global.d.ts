
// обязательная установка sass пакета (npm i sass --save-dev)

declare module '*.module.scss' {
  const css: Record<string, string>;
  export default css;
}

declare module '*.scss' {
  const css: Record<string, string>;
  export default css;
}