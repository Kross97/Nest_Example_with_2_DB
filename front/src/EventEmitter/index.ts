interface IHandler {
  once: boolean;
  handler: () => void;
}

export const EventsDictionary = {
  unAuthorized: "unAuthorized",
} as const;

type TEvents = keyof typeof EventsDictionary;

class EventEmitter {
  private eventsHandlersDispatcher: Record<string, IHandler[]> = {};

  add(event: TEvents, handler: () => void, once?: boolean) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event].push({ handler, once: !!once });
    } else {
      this.eventsHandlersDispatcher[event] = [{ handler, once: !!once }];
    }
  }

  remove(event: TEvents, handler: () => void) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => handlerObj.handler !== handler
      );
    }
  }

  emit(event: TEvents) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => {
          handlerObj.handler();
          return !handlerObj.once;
        }
      );
    }
  }
}

export const EventEmitterAgent = new EventEmitter();
