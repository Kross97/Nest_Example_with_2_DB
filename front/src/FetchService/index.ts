import {  getTokenJwt } from '../provider/AllProvider';
import { errorHandler } from './utils/errorHandler';

interface IUrlParams {
  url: string;
  headers?: Record<string, string>
  body?: any;
  requestType?: 'none' | 'json';
  responseType?: 'json' | 'blob' | 'text' | 'arrayBuffer';
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
}

const dispatcherHeaders = {
  'json': { 'content-type': 'application/json; charset=utf-8' },
  'none': {}
};

const dispatcherBody = {
  'json': (body: any) => JSON.stringify(body),
  'none': (body: any) => body,
}

const initAuthHeader = () => ({ 'Authorization': `Bearer ${getTokenJwt()}`})

class FetchService {

  private backUrl: string = process.env.BACK_HOST || 'http://localhost:3001';

  async request<T = any>({ url, headers = {}, body = '', method, requestType = 'none', responseType = 'json' }: IUrlParams) {
    const urlHost = (this.backUrl + (url.startsWith('/') ? url : `/${url}`));
    const currentHeaders = {...headers, ...dispatcherHeaders[requestType], ...initAuthHeader()};
    const currentUrl = method === 'GET' ? `${urlHost}?${new URLSearchParams(body).toString()}` : urlHost;

    const response = await fetch(currentUrl, {
      headers: currentHeaders,
      method,
      body: method === 'GET' ? null : dispatcherBody[requestType](body),
      credentials: 'include'
    });

    if(response.ok) {
      const contentLength = Number(response.headers.get('content-length') || 0);
      if (contentLength) {
        return response[responseType]();
      }
    } else {
      errorHandler(response);
    }
  }

  async postRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'POST' });
  }

  async getRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'GET' });
  }

  async putRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'PUT' });
  }
};

export const FetchAgent = new FetchService();