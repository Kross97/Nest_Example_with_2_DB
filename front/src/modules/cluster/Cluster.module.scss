.clusterBlock {
  display: flex;
  flex-direction: column;
  row-gap: 14px;
  border: 1px solid gray;
  background-color: #d97436;
  padding: 20px;
  width: 100%;
}

.clusterBlock h2, .clusterBlock h3 {
  margin: 0;
}