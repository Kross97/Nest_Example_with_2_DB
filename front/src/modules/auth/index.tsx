import cn from './Auth.module.scss';
import React, { ChangeEvent,  useState } from 'react';
import { FetchAgent } from '../../FetchService';
import { useAllContext } from '../../provider/AllProvider';


export const AuthBlock = () => {
  const [data, setUser] = useState<Record<string, string>>({});

  const { setToken, setCurrentPortHandler } = useAllContext();

  const change = (e: ChangeEvent<HTMLInputElement>) => {
    const nameField = e.target.name;
    const value = e.target.value;
    setUser((prev) => ({ ...prev, [nameField]: value }));
  };

  const signIn = async () => {
     const tokenData = await FetchAgent.postRequest<{ token: string; currentPort: number}>({ url: 'auth/signIn', requestType: 'json', body: data });
     console.log('tokenData =>', tokenData)
     setToken(tokenData.token);
     setCurrentPortHandler(tokenData.currentPort);
  };

  return <div className={cn.authBlock}>
    <h3>Авторизация</h3>
    <span>Логин</span>
    <input onChange={change} type={'text'} name={'login'} value={data.login}/>
    <span>Пароль</span>
    <input onChange={change} type={'password'} name={'password'} value={data.password}/>
    <button onClick={signIn}>Войти</button>
  </div>;
};