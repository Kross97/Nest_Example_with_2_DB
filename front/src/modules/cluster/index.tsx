import React, { useEffect, useState } from "react";
import cn from "./Cluster.module.scss";
import { FetchAgent } from "../../FetchService";
import { IPortsData, useAllContext } from "../../provider/AllProvider";

export function ClusterBlock() {
  const [currentPorts, setPorts] = useState<number[]>([]);

  const { setCurrentPortHandler, currentPort } = useAllContext();

  const getCurrentPorts = () => {
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_ports" })
      .then((results: IPortsData) => {
        setPorts(results.availablePorts);
      })
      .catch((err) => {
        console.log("Ошибка получения портов", Object.entries(err));
      });
  };

  useEffect(() => {
    getCurrentPorts();
  }, []);

  const enableClusterHandle = () =>
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_enable" }).then(() => {
      getCurrentPorts();
    });

  const disableClusterHandler = () => {
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_exit" }).then(() => {
      getCurrentPorts();
      setCurrentPortHandler();
    });
  };

  const setNewPortServer = () => {
    void setCurrentPortHandler();
  };

  return (
    <div className={cn.clusterBlock}>
      <h2>Работа с кластеризацией сервера (Cluster модуль node)</h2>
      <h3>{`Состояние кластеризации: ${currentPorts.length === 1 ? "Отключена" : "Включена"}`}</h3>
      <h3>{`Текущий порт: ${currentPort}`}</h3>
      <div>
        <span>Доступные порты:</span>
        <span>
          {currentPorts.map((currentPort) => (
            <React.Fragment key={currentPort}>
              <span style={{ cursor: "pointer", color: "blue" }} onClick={() => setCurrentPortHandler(currentPort)}>
                {currentPort}
              </span>
              <span>{` , `}</span>
            </React.Fragment>
          ))}
        </span>
      </div>
      <div>
        <button disabled={currentPorts.length === 1} onClick={setNewPortServer}>
          Выбрать новый порт сервера
        </button>
        <button disabled={currentPorts.length !== 1} onClick={enableClusterHandle}>
          Включить кластеризацию
        </button>
        <button disabled={currentPorts.length === 1} onClick={disableClusterHandler}>
          Отключить кластеризацию
        </button>
      </div>
    </div>
  );
}
