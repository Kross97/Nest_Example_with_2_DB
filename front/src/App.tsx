import React from 'react';
import './App.css';
import { PhotoBlock } from './modules/photos';
import { UserBlock } from './modules/user';
import { AuthBlock } from './modules/auth';
import {  useAllContext } from './provider/AllProvider';
import { CryptoBlock } from './modules/crypto';

function App() {
  const { token, setToken } = useAllContext();

  const exitHandler = () => {
    setToken(null);
  };

  return (
    <>
      {token && <button onClick={exitHandler} style={{ position: 'fixed', top: '50px', right: '50px'}}>exit</button>}
      <div style={{ display: 'flex'}}>
        {token && <PhotoBlock />}
        {token && <UserBlock />}
        {token && <CryptoBlock />}
        {!token && <AuthBlock />}
      </div>
    </>
  );
}

export default App;
