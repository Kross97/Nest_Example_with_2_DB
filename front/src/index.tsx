import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import { FullApp } from './App';
import { AllProvider } from './provider/AllProvider';
import './FetchService/index';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <AllProvider>
      <FullApp />
    </AllProvider>
  </React.StrictMode>
);
