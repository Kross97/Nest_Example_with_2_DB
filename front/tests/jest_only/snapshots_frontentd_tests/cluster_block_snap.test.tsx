//@ts-nocheck

/**
 * Первый способ переделать внутреннюю логику метода
 * */
// jest.mock('../../src/FetchService', () => {
//   const { FetchAgent } = jest.requireActual('../../src/FetchService');
//   return {
//     FetchAgent: {
//       ...FetchAgent,
//       getRequest: jest.fn(() => Promise.resolve({
//         availablePorts: [3002, 3003],
//         currentPort: 3001,
//       }))
//     }
//   }
// });


import { ClusterBlock } from '../../../src/modules/cluster'
import { create, act } from 'react-test-renderer';
import { FetchAgent } from '../../../src/FetchService';
import React from 'react';
import Mock = jest.Mock;


/**
 * Второй способ переписать внутренню логику метода
 * */


const getRequestMock = jest.spyOn<typeof FetchAgent, 'getRequest', undefined>(FetchAgent, 'getRequest').mockImplementation(() => Promise.resolve({
  availablePorts: [3002, 3003],
  currentPort: 3001,
}))


/**
 * Делаем имитацию useState, чтобы не было пере-рендера в компоненте из за useEffecta
 * */
const setStateMock = jest.fn();

/**
 * Моковые данные для корректного снапшота с данными
 * */
const useSateMock = () => [[3002,3003], setStateMock];
const reactUseState = jest.spyOn<typeof React, 'useState', undefined>(React, "useState").mockImplementation(useSateMock);

afterAll(() => {
  /**
   * Возвращаем исходные реализации методов
   * */
  getRequestMock.mockRestore();
  reactUseState.mockRestore();
});

describe.skip('Тестирование снапшотами блока Cluster', () => {
  test('Тест первый',  () => {
        const mockGetRequest = FetchAgent.getRequest as Mock;

        let tree;
         act(() => {
          tree = create(<ClusterBlock />);
        });

       expect(FetchAgent.getRequest).toHaveBeenCalled();
       expect(mockGetRequest.mock.results[0].value).resolves.toEqual({ availablePorts: [ 3002, 3003 ], currentPort: 3001 })
       expect(tree.toJSON()).toMatchSnapshot();
  });
});