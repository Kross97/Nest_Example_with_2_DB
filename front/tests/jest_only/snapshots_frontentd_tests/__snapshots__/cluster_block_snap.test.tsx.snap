// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование снапшотами блока Cluster Тест первый 1`] = `
<div
  className="clusterBlock"
>
  <h2>
    Работа с кластеризацией сервера (Cluster модуль node)
  </h2>
  <h3>
    Состояние кластеризации: Включена
  </h3>
  <h3>
    Текущий порт: 0
  </h3>
  <div>
    <span>
      Доступные порты:
    </span>
    <span>
      <span
        onClick={[Function]}
        style={
          {
            "color": "blue",
            "cursor": "pointer",
          }
        }
      >
        3002
      </span>
      <span>
         , 
      </span>
      <span
        onClick={[Function]}
        style={
          {
            "color": "blue",
            "cursor": "pointer",
          }
        }
      >
        3003
      </span>
      <span>
         , 
      </span>
    </span>
  </div>
  <div>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Выбрать новый порт сервера
    </button>
    <button
      disabled={true}
      onClick={[Function]}
    >
      Включить кластеризацию
    </button>
    <button
      disabled={false}
      onClick={[Function]}
    >
      Отключить кластеризацию
    </button>
  </div>
</div>
`;
